src/12_graphics_pipeline_complete.rs

```diff
@@ -0,0 +1,510 @@
+            },
+            hal::pso::EntryPoint::<back::Backend> {
+                entry: "main",
+                module: &frag_shader_module,
+                specialization: &[],
+            },
+        );
+
+        // shader stage set up
+        let shaders = hal::pso::GraphicsShaderSet {
+            vertex: vs_entry,
+            hull: None,
+            domain: None,
+            geometry: None,
+            fragment: Some(fs_entry),
+        };
+
+        // rasterizer: is there a field to set the rasterizer discard option?
+        let rasterizer = hal::pso::Rasterizer {
+            depth_clamping: false,
+            polygon_mode: hal::pso::PolygonMode::Fill,
+            cull_face: <hal::pso::Face>::BACK,
+            front_face: hal::pso::FrontFace::Clockwise,
+            depth_bias: None,
+            conservative: false,
+        };
+
+        // no need to set up vertex input format, as it is hardcoded
+        let vertex_buffers: Vec<hal::pso::VertexBufferDesc> = Vec::new();
+        let attributes: Vec<hal::pso::AttributeDesc> = Vec::new();
+
+        // input assembly
+        let input_assembler = hal::pso::InputAssemblerDesc::new(hal::Primitive::TriangleList);
+
+        // implements optional blending description provided in vulkan-tutorial
+        let blender = {
+            let blend_state = hal::pso::BlendState::On {
+                color: hal::pso::BlendOp::Add {
+                    src: hal::pso::Factor::One,
+                    dst: hal::pso::Factor::Zero,
+                },
+                alpha: hal::pso::BlendOp::Add {
+                    src: hal::pso::Factor::One,
+                    dst: hal::pso::Factor::Zero,
+                },
+            };
+
+            hal::pso::BlendDesc {
+                logic_op: Some(hal::pso::LogicOp::Copy),
+                targets: vec![hal::pso::ColorBlendDesc(
+                    hal::pso::ColorMask::ALL,
+                    blend_state,
+                )],
+            }
+        };
+
+        let depth_stencil = hal::pso::DepthStencilDesc {
+            depth: hal::pso::DepthTest::Off,
+            depth_bounds: false,
+            stencil: hal::pso::StencilTest::Off,
+        };
+
+        let multisampling: Option<hal::pso::Multisampling> = None;
+
+        // viewports and scissors
+        let baked_states = hal::pso::BakedStates {
+            viewport: Some(hal::pso::Viewport {
+                rect: hal::pso::Rect {
+                    x: 0,
+                    y: 0,
+                    w: extent.width as i16,
+                    h: extent.width as i16,
+                },
+                depth: (0.0..1.0),
+            }),
+            scissor: Some(hal::pso::Rect {
+                x: 0,
+                y: 0,
+                w: extent.width as i16,
+                h: extent.height as i16,
+            }),
+            blend_color: None,
+            depth_bounds: None,
+        };
+
+        // dynamic state is handled automatically by HAL
```

Comment:
We can probably leave this out – dynamic state still needs to be specified to HAL (the caller chooses between static/dynamic types) but it doesn't seem like its needed in these examples